mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1009,10 +1009,17 @@ private[spark] object RandomForest extends Logging {
       // sort distinct values
       val valueCounts = valueCountMap.toSeq.sortBy(_._1).toArray
 
-      // if possible splits is not enough or just enough, just return all possible splits
       val possibleSplits = valueCounts.length - 1
-      if (possibleSplits <= numSplits) {
-        valueCounts.map(_._1).init
+      if (possibleSplits == 0) {
+        // constant feature
+        Array.empty[Double]
+      } else if (possibleSplits <= numSplits) {
+        // if possible splits is not enough or just enough, just return all possible splits
+        val splits = for {
+          i <- 0 until possibleSplits
+        } yield (valueCounts(i)._1 + valueCounts(i + 1)._1) / 2
```

Comment:
Good idea. Maybe even just `(0 until possibleSplits).map(...).toArray` which is probably about the same thing anyway. You might write `/ 2.0` to be clear it's floating point division